tower-discover/src/lib.rs

```diff
@@ -43,6 +44,24 @@ pub trait Discover {
     ) -> Poll<Result<Change<Self::Key, Self::Service>, Self::Error>>;
 }
 
+// delegate through Pin
+impl<P> Discover for Pin<P>
```

Comment:
I believe we need to add a &mut T and Box<T> impls as well. This is what Future and Stream do.